optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -739,3 +651,331 @@ def best_trial(self):
+                "distributed environment, please use network file system (e.g., NFS) to share "
+                "models with multiple workers.".format(self._model_dir)
+            )
+
+        with open(path, "rb") as fin:
+            booster = pickle.load(fin)
+
+        return booster
+
+    def tune_params(self, target_param_names, n_trials, sampler, step_name):
+        # type: (List[str], int, optuna.samplers.BaseSampler, str) -> OptunaObjective
+
+        objective = super(LightGBMTuner, self).tune_params(
+            target_param_names, n_trials, sampler, step_name
+        )
+
+        # Add tuning history.
+        self.tuning_history += objective.report
+
+        if objective.best_booster_with_trial_number is not None:
+            self._best_booster_with_trial_number = objective.best_booster_with_trial_number
+            self._best_params.update(self.best_params)
+        return objective
+
+    def _create_objective(
+        self,
+        target_param_names: List[str],
+        train_set: lgb.Dataset,
+        step_name: str,
+        pbar: tqdm.tqdm,
+    ) -> OptunaObjective:
+        return OptunaObjective(
+            target_param_names,
+            self.lgbm_params,
+            train_set,
+            self.lgbm_kwargs,
+            self.best_score,
+            step_name=step_name,
+            model_dir=self._model_dir,
+            pbar=pbar,
+        )
+
+
+class LightGBMTunerCV(LightGBMBaseTuner):
+    """Hyperparameter-tuning with Optuna for LightGBM.
+
+    In each trial, it performs cross-validation with the suggested hyperparameters.
+    Arguments and keyword arguments for `lightgbm.cv()
+    <https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.cv.html>`_ can be passed except
+    ``metrics``, ``init_model`` and ``eval_train_metric``.
+    The arguments that only :class:`~optuna.integration.lightgbm.LightGBMTunerCV` has are listed
+    below:
+
+    Args:
+        time_budget:
+            A time budget for parameter tuning in seconds.
+
+        study:
+            A :class:`~optuna.study.Study` instance to store optimization results. The
+            :class:`~optuna.trial.Trial` instances in it has the following user attributes:
+            ``elapsed_secs`` is the elapsed time since the optimization starts.
+            ``average_iteration_time`` is the average time of iteration to train the booster
+            model in the trial. ``lgbm_params`` is a JSON-serialized dictionary of LightGBM
+            parameters used in the trial.
+
+        optuna_callbacks:
+            List of Optuna callback functions that are invoked at the end of each trial.
+            Each function must accept two parameters with the following types in this order:
+            :class:`~optuna.study.Study` and :class:`~optuna.FrozenTrial`.
+            Please note that this is not a ``callbacks`` argument of `lightgbm.train()`_ .
+
```

Comment:
I didn't add `model_dir` because the original `lightgbm.cv` does not return any boosters. To save them, I think we need to implement CV with `lightgbm.train`. I think `model_dir` is out of the scope of this PR.
Does it make sense?

Reply:
I see. I totally agree with your claim. Thank you for your detailed explanation.